File: test/rbi_gen_package_runner.rb
# frozen_string_literal: true
# typed: true

require 'json'
require 'optparse'
require 'set'
require 'tempfile'

require 'sorbet-runtime'

class Module
  include T::Sig
end

class PackageInfo
  sig {returns(String)}
  attr_reader :name

  sig {returns(T::Array[String])}
  attr_reader :imports

  sig {returns(T::Array[String])}
  attr_reader :testImports

  sig {returns(T::Array[String])}
  attr_reader :files

  sig {returns(T::Array[String])}
  attr_reader :testFiles

  sig {returns(String)}
  attr_reader :mangled_name

  sig {params(hash: T::Hash[T.untyped, T.untyped]).returns(PackageInfo)}
  def self.from_hash(hash)
    self.new(hash.fetch("name"),
                    hash.fetch("imports"),
                    hash.fetch("testImports"),
                    hash.fetch("files"),
                    hash.fetch("testFiles"))
  end

  private_class_method :new

  sig {params(
         name: String,
         imports: T::Array[String],
         testImports: T::Array[String],
         files: T::Array[String],
         testFiles: T::Array[String]).void}
  def initialize(name, imports, testImports, files, testFiles)
    @name = name
    @imports = imports
    @testImports = testImports
    @files = files
    @testFiles = testFiles

    @mangled_name = T.let("#{name.gsub('::', '_')}_Package", String)
  end
end

class DependencyInfo < T::Struct
  const :packageRefs, T::Array[String]
  const :rbiRefs, T::Array[String]

  def self.load(file)
    from_hash(JSON.parse(File.read(file)))
  end
end

class PackageDB
  sig {returns(T::Hash[String, String])}
  attr_reader :rbis

  sig {returns(T::Hash[String, DependencyInfo])}
  attr_reader :deps

  sig {returns(T::Hash[String, PackageInfo])}
  attr_reader :info

  sig {params(package_dir: String, package_info: T::Array[PackageInfo]).void}
  def initialize(package_dir, package_info)
    @rbis = T.let({}, T::Hash[String, String])
    @deps = T.let({}, T::Hash[String, DependencyInfo])
    @info = T.let(package_info.to_h {|info| [info.name, info]}, T::Hash[String, PackageInfo])

    Dir.new(package_dir).each do |entry|
      next if entry == '.' or entry == '..'

      case
      when entry =~ /([^.]+).deps.json$/
        @deps[$1] = DependencyInfo.load(File.join(package_dir, entry))
      when entry =~ /([^.]+).package.rbi$/
        @rbis[$1] = entry
      else
        raise "Unknown filename in package directory #{entry}"
      end
    end
  end

  sig {params(info: PackageInfo).returns(String)}
  def lookup_rbi_for(info)
    rbis.fetch(info.mangled_name)
  end
  
  sig {params(info: PackageInfo).returns(DependencyInfo)}
  def lookup_deps_for(info)
    deps.fetch(info.mangled_name)
  end

  sig {params(name: String).returns(PackageInfo)}
  def lookup_package_for(name)
    info.fetch(name)
  end
end

module Runner
  include T::Sig

  sig {params(command: T::Array[String], chdir: T.nilable(String)).void}
  def self.check_call(command, chdir: nil)
    opts = {}
    if chdir
      opts[:chdir] = chdir
    end
    pid = T.unsafe(Process).spawn({}, *command, opts)
    Process.wait(pid)
    status = $?
    if status.to_i != 0
      raise "#{command.join(" ")} exited with #{status.to_i}"
    end
  end

  sig {params(package_info_file: Tempfile).returns(T::Array[PackageInfo])}
  def self.load_package_info(package_info_file)
    json = JSON.parse(File.read(package_info_file))

    json.map do |package|
      p = PackageInfo.from_hash(package)
    end
  end

  sig {params(sorbet: String, root: String, test_directory: String,
              required_files: T::Set[String], rbi_package_dir: String,
              rbi_files: T::Set[String]).void}
  def self.verify_single_package_typechecking(
        sorbet, root, test_directory, required_files,
        rbi_package_dir, rbi_files)
    Dir.mktmpdir do |tmpdir|
      # Copy the package's Ruby files to the tmpdir, mirroring the directory
      # structure.
      dirnames = required_files.map {|f| File.dirname(f)}.uniq
      dirnames.each do |dirname|
        target = File.join(tmpdir, dirname)
        FileUtils.mkdir_p(target)
      end

      Dir.chdir(root) do |root|
        required_files.each do |f|
          dirname = File.dirname(f)
          target = File.join(tmpdir, dirname)
          FileUtils.cp(f, target)
        end
      end

      # Copy the RBI files into the root.
      rbi_files.each do |rbi|
        source = File.join(rbi_package_dir, rbi)
        FileUtils.cp(source, File.join(tmpdir, test_directory))
      end

      check_call([
                   sorbet,
                   '--silence-dev-message',
                   '--stripe-mode',
                   '--ignore',
                   '__package.rb',
                   test_directory
                 ],
                 chdir: tmpdir)
    end
  end

  sig {params(sorbet: String, root: String, test_directory: String).void}
  def self.main(sorbet:, root:, test_directory:)
    Tempfile.open('package_info') do |package_info_file|
      check_call([
                   sorbet,
                   '--silence-dev-message',
                   '--stripe-mode',
                   '--stripe-packages',
                   '--dump-package-info',
                   T.must(package_info_file.path),
                   test_directory
                 ], chdir: root)
      
      package_info = load_package_info(package_info_file)

      package_directories = package_info.map do |package|
        dirs = package.files.map {|f| File.dirname(f)}.uniq
        if dirs.size != 1 
          raise "Package #{package.name} must be limited to a single directory"
        end

        dirs[0]
      end
      
      Dir.mktmpdir do |rbi_package_dir|
        check_call([
                     sorbet,
                     '--silence-dev-message',
                     '--stripe-mode',
                     '--package-rbi-generation',
                     '--package-rbi-dir',
                     rbi_package_dir,
                     '--ignore',
                     '__package.rb',
                     test_directory],
                   chdir: root)

        db = PackageDB.new(rbi_package_dir, package_info)

        package_info.map do |info|
          required_files = T.let(Set.new, T::Set[String])
          rbi_files = T.let(Set.new, T::Set[String])
          visited = T.let(Set.new, T::Set[String])
          worklist = T.let([], T::Array[String])

          # We always need the rb files from the package itself, but we should only
          # need the RBI files from any transitive imports.
          required_files.merge(info.files)
          required_files.merge(info.testFiles)
          visited.add(info.name)
          worklist.append(*info.imports)
          worklist.append(*info.testImports)

          while !worklist.empty?
            pkg = T.must(worklist.pop)
            next if visited.include?(pkg)

            visited.add(pkg)
            info_for_dependency = db.lookup_package_for(pkg)

            rbi_files.add(db.lookup_rbi_for(info_for_dependency))

            worklist.append(*info_for_dependency.imports)
            worklist.append(*info_for_dependency.testImports)
          end

          verify_single_package_typechecking(sorbet, root, test_directory, required_files,
                                             rbi_package_dir, rbi_files)
        end
      end
    end
  end
end

if __FILE__ == $0
  sorbet = T.let(nil, T.nilable(String))
  root = T.let(nil, T.nilable(String))
  test_directory = T.let(nil, T.nilable(String))

  OptionParser.new do |opts|
    opts.on '--sorbet=PATH', 'Path to the Sorbet executable' do |path|
      sorbet = File.realpath(path)
    end

    opts.on '--root=PATH', 'Path to the root of the Sorbet checkout' do |path|
      root = File.realpath(path)
    end

    opts.on '--test-directory=PATH', 'Relative path to the root of the checked directory' do |path|
      test_directory = path
    end
  end.parse!

  Runner.main(
    sorbet: T.must(sorbet),
    root: T.must(root),
    test_directory: T.must(test_directory)
  )
end

File: test/rbi_gen_package_runner.cc
#include "absl/strings/match.h"
#include "absl/strings/str_replace.h"
#include "common/Subprocess.h"
#include "common/common.h"
#include "rapidjson/document.h"
#include "rapidjson/istreamwrapper.h"
#include <cstdlib>
#include <deque>
#include <filesystem>
#include <fstream>
#include <memory>
#include <string>
#include <unistd.h>
#include <vector>

using namespace std;

namespace {

using defer = shared_ptr<void>;

vector<string> fromJsonStringArray(const rapidjson::Value &elems) {
    vector<string> result;
    for (const auto &elem : elems.GetArray()) {
        result.emplace_back(elem.GetString());
    }

    return result;
}

rapidjson::Document parseJsonFile(string filename) {
    ifstream ifs(filename);
    rapidjson::IStreamWrapper isw(ifs);
    rapidjson::Document document;
    document.ParseStream(isw);
    if (document.HasParseError()) {
        throw new runtime_error("Failed to parse document: " + filename);
    }

    return document;
}

struct PackageInfo {
    string name;
    vector<string> imports;
    vector<string> testImports;
    vector<string> files;
    vector<string> testFiles;
    string mangledName;

    PackageInfo() = default;
    PackageInfo(string name, vector<string> imports, vector<string> testImports, vector<string> files,
                vector<string> testFiles)
        : name(name), imports(imports), testImports(testImports), files(files), testFiles(testFiles),
          mangledName(absl::StrReplaceAll(name, {{"::", "_"}}) + "_Package"s) {}

    static PackageInfo fromJson(const rapidjson::Value &d) {
        return PackageInfo(d["name"].GetString(), fromJsonStringArray(d["imports"]),
                           fromJsonStringArray(d["testImports"]), fromJsonStringArray(d["files"]),
                           fromJsonStringArray(d["testFiles"]));
    }
};

struct DependencyInfo {
    vector<string> packageRefs;
    vector<string> rbiRefs;

    DependencyInfo() = default;
    DependencyInfo(vector<string> packageRefs, vector<string> rbiRefs) : packageRefs(packageRefs), rbiRefs(rbiRefs) {}

    static DependencyInfo load(string file) {
        auto d = parseJsonFile(file);
        return DependencyInfo(fromJsonStringArray(d["packageRefs"]), fromJsonStringArray(d["rbiRefs"]));
    }
};

absl::flat_hash_map<string, PackageInfo> packageInfoToHash(vector<PackageInfo> packageInfo) {
    absl::flat_hash_map<string, PackageInfo> result;
    for (auto &info : packageInfo) {
        result[info.name] = info;
    }

    return result;
}

struct PackageDB {
    absl::flat_hash_map<string, string> rbis;
    absl::flat_hash_map<string, DependencyInfo> deps;
    absl::flat_hash_map<string, PackageInfo> info;

    PackageDB(string packageDir, vector<PackageInfo> packageInfo)
        : rbis({}), deps({}), info(packageInfoToHash(packageInfo)) {
        for (const auto &fullEntry : filesystem::directory_iterator(filesystem::path(packageDir))) {
            const auto &entry = fullEntry.path().filename();
            if (entry == "." || entry == "..") {
                continue;
            }

            auto filepath = entry.native();
            if (absl::EndsWith(filepath, ".deps.json")) {
                deps[filepath.substr(0, filepath.size() - 10)] = DependencyInfo::load(packageDir + "/" + filepath);
            } else if (absl::EndsWith(filepath, ".package.rbi")) {
                rbis[filepath.substr(0, filepath.size() - 12)] = filepath;
            } else {
                throw new runtime_error("Unknown filename in package directory " + filepath);
            }
        }
    }

    string lookupRbiFor(PackageInfo info) {
        return rbis.at(info.mangledName);
    }

    DependencyInfo lookupDepsFor(PackageInfo info) {
        return deps.at(info.mangledName);
    }

    PackageInfo lookupPackageFor(string name) {
        return info.at(name);
    }
};

string copyFromOwnedCharStar(char *str) {
    string result(str);
    free(str);
    return result;
}

string mktempTemplate(string basename) {
    string packageInfoFilename = basename + ".XXXXXXXXXX";
    if (const char *envTmpdir = getenv("TMPDIR")) {
        packageInfoFilename = string(envTmpdir) + "/" + packageInfoFilename;
    } else {
        packageInfoFilename = "/tmp/" + packageInfoFilename;
    }

    return packageInfoFilename;
}

void chDirAndSpawn(string executable, vector<string> arguments, optional<string_view> stdinContents,
                   optional<string> chdir) {
    auto pwd = filesystem::current_path();
    defer _(nullptr, [&](...) { filesystem::current_path(pwd); });
    if (chdir.has_value()) {
        filesystem::path chdirPath = chdir.value();
        filesystem::current_path(chdirPath);
    }
    auto result = sorbet::Subprocess::spawn(move(executable), move(arguments), nullopt);

    if (!result.has_value() || result->status != 0) {
        throw runtime_error("Subprocess failed");
    }
}

vector<PackageInfo> loadPackageInfo(string packageInfoFilename) {
    auto document = parseJsonFile(packageInfoFilename);

    vector<PackageInfo> result;
    for (const auto &package : document.GetArray()) {
        result.emplace_back(PackageInfo::fromJson(package));
    }

    return result;
}

void verifySinglePackageTypechecking(string sorbet, string root, string testDirectory,
                                     absl::flat_hash_set<string> requiredFiles, string rbiPackageDir,
                                     absl::flat_hash_set<string> rbiFiles) {
    string tmpdirFilename = mktempTemplate("tmp");
    auto tmpdir = mkdtemp(tmpdirFilename.data());
    filesystem::path tmpdirPath = tmpdir;

    absl::flat_hash_set<string> dirnames;
    for (const auto &f : requiredFiles) {
        dirnames.emplace(filesystem::path(f).parent_path());
    }
    for (const auto &dirname : dirnames) {
        filesystem::path target = tmpdirPath / filesystem::path(dirname);
        filesystem::create_directories(target);
    }

    auto pwd = filesystem::current_path();
    filesystem::current_path(filesystem::path(root));
    for (const auto &f : requiredFiles) {
        auto fPath = filesystem::path(f);
        auto dirname = fPath.parent_path();
        // For some reason, I couldn't get the filesystem::copy_file call to work, and I think it
        // has something to do with bazel sandboxing. I didn't look into it further.
        chDirAndSpawn("cp",
                      {
                          fPath.native(),
                          (tmpdirPath / dirname).native(),
                      },
                      nullopt, nullopt);
    }
    filesystem::current_path(pwd);

    for (const auto &rbi : rbiFiles) {
        auto source = filesystem::path(rbiPackageDir) / filesystem::path(rbi);
        chDirAndSpawn("cp",
                      {
                          source.native(),
                          (filesystem::path(tmpdirPath) / filesystem::path(testDirectory)).native(),
                      },
                      nullopt, nullopt);
    }

    chDirAndSpawn(sorbet,
                  {
                      "--silence-dev-message",
                      "--stripe-mode",
                      "--ignore",
                      "__package.rb",
                      testDirectory,
                  },
                  nullopt, tmpdir);
}

} // namespace

int main(int argc, char **argv) {
    if (argc != 4) {
        return 1;
    }

    auto sorbet = copyFromOwnedCharStar(realpath(argv[1], nullptr));
    auto root = copyFromOwnedCharStar(realpath(argv[2], nullptr));
    auto testDirectory = string(argv[3]);

    vector<PackageInfo> packageInfo;
    vector<string> packageDirectories;
    {
        string packageInfoFilename = mktempTemplate("package_info");
        auto packageInfoFd = mkstemp(packageInfoFilename.data());
        if (packageInfoFd == -1) {
            return 1;
        }
        defer _(nullptr, [&](...) { unlink(packageInfoFilename.c_str()); });
        close(packageInfoFd);

        chDirAndSpawn(sorbet,
                      {
                          "--silence-dev-message"s,
                          "--stripe-mode"s,
                          "--stripe-packages"s,
                          "--dump-package-info"s,
                          packageInfoFilename,
                          testDirectory,
                      },
                      nullopt, root);

        packageInfo = loadPackageInfo(packageInfoFilename);
        for (const auto &package : packageInfo) {
            absl::flat_hash_set<string> dirs;
            for (const auto &f : package.files) {
                dirs.emplace(filesystem::path(f).parent_path());
            }

            if (dirs.size() != 1) {
                throw runtime_error("Package must be limited to a single directory");
            }

            packageDirectories.emplace_back(*dirs.begin());
        }
    }

    {
        string rbiPackageDirFilename = mktempTemplate("tmp");
        auto rbiPackageDir = mkdtemp(rbiPackageDirFilename.data());
        chDirAndSpawn(sorbet,
                      {
                          "--silence-dev-message",
                          "--stripe-mode",
                          "--package-rbi-generation",
                          "--package-rbi-dir",
                          rbiPackageDir,
                          "--ignore",
                          "__package.rb",
                          testDirectory,
                      },
                      nullopt, root);

        PackageDB db(rbiPackageDir, packageInfo);

        for (const auto &info : packageInfo) {
            absl::flat_hash_set<string> requiredFiles;
            absl::flat_hash_set<string> rbiFiles;
            absl::flat_hash_set<string> visited;
            deque<string> worklist;

            for (const auto &file : info.files) {
                requiredFiles.emplace(file);
            }
            for (const auto &file : info.testFiles) {
                requiredFiles.emplace(file);
            }
            visited.emplace(info.name);
            for (const auto &import_ : info.imports) {
                worklist.emplace_back(import_);
            }
            for (const auto &import_ : info.testImports) {
                worklist.emplace_back(import_);
            }

            while (!worklist.empty()) {
                auto pkg = worklist.front();
                worklist.pop_front();
                if (visited.count(pkg) == 1) {
                    continue;
                }

                visited.emplace(pkg);
                auto infoForDependency = db.lookupPackageFor(pkg);
                rbiFiles.emplace(db.lookupRbiFor(infoForDependency));
                for (const auto &import_ : infoForDependency.imports) {
                    worklist.emplace_back(import_);
                }
                for (const auto &import_ : infoForDependency.testImports) {
                    worklist.emplace_back(import_);
                }
            }

            verifySinglePackageTypechecking(sorbet, root, testDirectory, requiredFiles, rbiPackageDir, rbiFiles);
        }
    }

    return 0;
}
